include/envoy/config/config_provider.h

```diff
@@ -41,6 +42,23 @@ class ConfigProvider {
   };
   using ConfigConstSharedPtr = std::shared_ptr<const Config>;
 
+  /**
+   * The type of API represented by a ConfigProvider.
+   */
+  enum class ApiType {
+    /**
+     * A "Full" API delivers a complete configuration as part of each resource (top level
+     * config proto); i.e., each resource contains the whole representation of the config intent. An
+     * example of this type of API is RDS.
+     */
+    Full,
+    /**
+     * A "Delta" API delivers a subset of the config intent as part of each resource (top level
+     * config proto). Examples of this type of API are CDS, LDS and SRDS.
+     */
+    Delta
```

Comment:
One very high level comment to kick off this review. I'm trying to puzzle out why delta/full needs to be special; why can't we treat full the same as a large delta update? My main goal is to reduce the amount of special purpose code we have for state-of-the-world vs. delta when looking at Envoy's config ingest architecture, so any divergence, while understandable in some cases, needs to have a good reason.

Reply:
I fully support the goal of minimizing divergence. The reason this API exposes slightly different interfaces is that the logic to handle "Full" API config updates is much simpler (due to not needing to maintain state between resources) and therefore the framework is able to offload/generalize more of that logic compared to "Delta" API updates.

Also, for clarification since the naming appears to be causing confusion, SRDS will be implemented as a "Delta" API in this framework even without initially supporting the DeltaDiscoveryRequest. The type is meant to convey that the API's top level config protos represent a delta update of the full config set, _even if the underlying subscription is not receiving delta updates_. The goal is to abstract the API's implementation from the underlying subscription details by simply conveying to the framework the semantics of the API.

Lastly, it _should_ be possible to generalize delta updates to a greater extent than is done now to minimize the differences; I am approaching this with a goal of iterating on the framework API as other DS APIs are implemented/ported so that we can progress towards that end goal.

Reply:
@fredlas 
> I think I have an even higher level question: why do state deltas get involved at all within ConfigProvider?

A goal of this framework is to share much of the common processing that needs to be done when implementing DS APIs to simplify development. For example, SRDS will mimic CDS' diffing of resource updates to determine when global and thread local data structures need to be updated with additions/modifications/deletes. I suspect future APIs with similar design constraints will require a similar mechanism, and therefore centralizing this logic into the framework seems appropriate to me. 

@htuch 
> Will the Delta path have essentially the same capabilities as Full, but with support for N config fragments rather than 1? How far do you plan on taking this in the SRDS work?

Yes, this is the end state that I'm aiming for. Right now, I'm trying to carefully balance complexity in the API by trying not to overgeneralize it too early. I am planning to port the RDS and SDS APIs to this framework, and as part of that effort I expect to move towards convergence in the framework.

Reply:
Fair enough; as discussed offline, I'm good with this strategy so will review with that in mind.